src/front/glsl/ast.rs

```diff
@@ -134,19 +134,117 @@ pub enum HirExprKind {
     },
 }
 
+#[derive(Debug, Hash, PartialEq, Eq)]
+pub enum QualifierKey {
+    String(String),
+    Layout,
+}
+
 #[derive(Debug)]
-pub enum TypeQualifier {
-    StorageQualifier(StorageQualifier),
-    Interpolation(Interpolation),
-    Set(u32),
-    Binding(u32),
-    Location(u32),
-    WorkGroupSize(usize, u32),
-    Sampling(Sampling),
+pub enum QualifierValue {
+    None,
+    Uint(u32),
     Layout(StructLayout),
-    Precision(Precision),
-    EarlyFragmentTests,
-    StorageAccess(StorageAccess),
+}
+
+#[derive(Debug, Default)]
+pub struct TypeQualifiers {
+    pub span: Span,
+    pub storage: (StorageQualifier, Span),
+    pub interpolation: Option<(Interpolation, Span)>,
+    pub precision: Option<(Precision, Span)>,
+    pub sampling: Option<(Sampling, Span)>,
+    pub storage_acess: Option<(StorageAccess, Span)>,
+    pub layout_qualifiers: crate::FastHashMap<QualifierKey, (QualifierValue, Span)>,
+}
+
+impl TypeQualifiers {
+    /// Appends `errors` with errors for all unused qualifiers
+    pub fn unused_errors(&self, errors: &mut Vec<super::Error>) {
+        if let Some((_, meta)) = self.interpolation {
+            errors.push(super::Error {
+                kind: super::ErrorKind::SemanticError(
+                    "Interpolation qualifiers can only be used in in/out variables".into(),
+                ),
+                meta,
+            });
+        }
+
+        if let Some((_, meta)) = self.sampling {
+            errors.push(super::Error {
+                kind: super::ErrorKind::SemanticError(
+                    "Sampling qualifiers can only be used in in/out variables".into(),
+                ),
+                meta,
+            });
+        }
+
+        if let Some((_, meta)) = self.storage_acess {
+            errors.push(super::Error {
+                kind: super::ErrorKind::SemanticError(
+                    "Memory qualifiers can only be used in storage variables".into(),
+                ),
+                meta,
+            });
+        }
+
+        for &(_, meta) in self.layout_qualifiers.values() {
+            errors.push(super::Error {
+                kind: super::ErrorKind::SemanticError("Unexpected qualifier".into()),
+                meta,
+            });
+        }
+    }
+
+    /// Removes the layout qualifier with `name`, if it exists and adds an error if it isn't
+    /// a [`QualifierValue::Uint`]
+    pub fn uint_layout_qualifier(
+        &mut self,
+        name: &str,
+        errors: &mut Vec<super::Error>,
+    ) -> Option<u32> {
+        match self
+            .layout_qualifiers
+            .remove(&QualifierKey::String(String::from(name)))
```

Comment:
Can `QualifierKey` have `Cow<>` instead of `String` in it to avoid heap allocations in places like this?